### ISSUE TITLE
Change the cos directory layout for pipeline artifacts

### ISSUE THREAD

**Issue by ptitzler** [elyra-ai/elyra#1745]
**Is your feature request related to a problem? Please describe.**
Spawned from https://github.com/elyra-ai/elyra/pull/1732

**Describe the solution you'd like**

Currently Elyra uses the following directory layout for input and output artifacts that are stored in a cloud storage bucket:
```
<pipeline-name-with-timestamp>/
<pipeline-name-with-timestamp>/<node-1-archive>.tgz    # input artifacts for node 1
<pipeline-name-with-timestamp>/<node-2-archive>.tgz    # input artifacts for node 2
<pipeline-name-with-timestamp>/<output-artifact-1>
<pipeline-name-with-timestamp>/<output-artifact-2>
<pipeline-name-with-timestamp>/<output-artifact-3>
...
```

There are two issues with this layout:
- input and output artifacts are stored in the same location
- output artifacts can be overwritten if the pipeline is run multiple times outside of Elyra (e.g. on KFP or AA)

To resolve the issues we could
- separate input and output artifacts
- associate output artifacts with a "run id" 

For example:

```
<pipeline-name-with-timestamp>/
<pipeline-name-with-timestamp>/<node-1-archive>.tgz
<pipeline-name-with-timestamp>/<node-2-archive>.tgz
<pipeline-name-with-timestamp>/<run-1-id>/<output-artifact-1>
<pipeline-name-with-timestamp>/<run-1-id>/<output-artifact-2>
<pipeline-name-with-timestamp>/<run-1-id>/<output-artifact-3>
<pipeline-name-with-timestamp>/<run-2-id>/<output-artifact-1>
<pipeline-name-with-timestamp>/<run-2-id>/<output-artifact-2>
<pipeline-name-with-timestamp>/<run-2-id>/<output-artifact-3>
...
```

Ideally this unique "run id" is something that's meaningful to the user.

**Comment by ptitzler** [comment]
This is an example how this could look for KFP:
- the top screen capture is from the KFP GUI, displaying pipeline runs
- the bottom screen capture is the object storage bucket content for above pipeline (note that only the marked run took advantage of a POC implementation for a proposal that utilizes the user-provided run name as "run id")

![image](https://user-images.githubusercontent.com/13068832/121599324-bd010d80-c9f7-11eb-8721-c51e672564c5.png)

**Comment by kevin-bates** [comment]
One of the items mentioned by @simon-streamotion in https://github.com/elyra-ai/elyra/discussions/1686 was from [this comment](https://github.com/elyra-ai/elyra/discussions/1686?sort=old#discussioncomment-778807):

> IMHO it would be cleaner to allow users to specify an additional s3 prefix so the '<pipeline_name-timestamp>' prefix is not created directly under the bucket. For example:
s3://my-bucket/my_project/mypipeline_0524024006
>
> This would allow our org to have a single bucket for elyra pipelines, but better organise the output.

I think we should consider adding an optional "project" (or "workspace" or "scope") _pipeline property_ that when set results in the additional level.  This would better facilitate shared COS credentials and collaboration in general within larger organizations.

**Comment by ptitzler** [comment]
Ah yes. On the back-end this can be easily implemented because a key name prefix is already passed to https://github.com/elyra-ai/kfp-notebook/blob/master/etc/docker-scripts/bootstrapper.py#L565 and https://github.com/elyra-ai/airflow-notebook/blob/master/etc/docker-scripts/bootstrapper.py#L371 and can accomodate another [optional]component:
```
<"project">/<pipeline-name-with-timestamp>/...
```

Not sure though how a "project" would be exposed to the user in the UI since it's an abstract organizational unit that does not have a representation in the GUI. If it's just a property value in the pipeline file, how would a user know which pipelines belong to the project?

**Comment by kevin-bates** [comment]
RIght.  Perhaps it might be better to treat it purely as a _classifier_  or _scope_ comes to mind.  However, I don't think we want to entertain having to open the pipeline files to determine their 'scope'.  If we need that reverse lookup (which is natural) this might require more thought.

**Comment by ptitzler** [comment]
> I don't think we want to entertain having to open the pipeline files to determine their 'scope'. If we need that reverse lookup (which is natural) this might require more thought.

I agree. We can always improve this incrementally and start by separating input and output artifacts. While such a change would impact the user workflow a bit, it doesn't have any impact on any of the Elyra tooling/UI. A new release would be good timing to make such a change.
